kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -66,245 +82,410 @@ String filterDescriptor() {
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug("{}: Dispatching downstream {} request to filter{}: {}",
-                        ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (msg instanceof InternalRequestFrame<?> decodedFrame) {
+            // jump the queue, internal request must flow!
+            writeDecodedRequest(decodedFrame, promise);
+        }
+        else if (msg instanceof DecodedRequestFrame<?> decodedFrame) {
+            if (writeFuture.isDone()) {
+                writeFuture = writeDecodedRequest(decodedFrame, promise);
             }
-
-            var stage = invoker.onRequest(decodedFrame.apiKey(), decodedFrame.apiVersion(), decodedFrame.header(),
-                    decodedFrame.body(), filterContext);
-            if (stage == null) {
-                if (LOGGER.isWarnEnabled()) {
-                    LOGGER.warn("{}: Filter{} for {} request unexpectedly returned null. This is a coding error in the filter. Closing connection.",
-                            ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
-                }
-                filterContext.closeConnection();
-                return CompletableFuture.completedFuture(null);
+            else {
+                writeFuture = writeFuture.whenComplete((a, b) -> {
+                    if (ctx.channel().isOpen()) {
+                        writeDecodedRequest(decodedFrame, promise);
+                    }
+                });
             }
-            var future = stage.toCompletableFuture();
-            boolean defer = !future.isDone();
-            var maybeDeferred = defer ? handleDeferredStage(ctx, future, decodedFrame) : future;
-            var execute = executeWrite(ctx, decodedFrame, filterContext, maybeDeferred);
-            var maybeDeferredCompleted = defer ? handleDeferredWriteCompletion(ctx, execute) : execute;
-            return maybeDeferredCompleted.thenApply(filterResult -> null);
         }
         else {
-            if (!(msg instanceof OpaqueRequestFrame)
-                    && msg != Unpooled.EMPTY_BUFFER) {
+            if (!(msg instanceof OpaqueRequestFrame) && msg != Unpooled.EMPTY_BUFFER) {
                 // Unpooled.EMPTY_BUFFER is used by KafkaProxyFrontendHandler#closeOnFlush
-                // but otherwise we don't expect any other kind of message
-                LOGGER.warn("Unexpected message writing to upstream: {}", msg, new IllegalStateException());
+                // but, otherwise we don't expect any other kind of message
+                throw new IllegalStateException("Unexpected message writing to upstream: " + msg);
             }
             ctx.write(msg, promise);
+        }
+    }
+
+    private CompletableFuture<Void> readDecodedResponse(DecodedResponseFrame<?> decodedFrame) {
```

Comment:
I'm not sure about the method name and I think that leads to it not having a single responsibility. 

Personally I think this is actually `dispatchDecodedResponse` and it potentially splits into `invokeFilterOnResponse` which covers L150 to L165. The result of that should be handed off to something else `forwardFilterResponse` covering the future management?  

I'm not wedded to any of the naming suggestions but I think this deserves to be broken down further.

Reply:
https://github.com/k-wall/kroxylicious/pull/8 looked good. I've incorporated into this PR.